rmw_fastrtps_shared_cpp/src/rmw_node.cpp

```diff
@@ -42,45 +47,60 @@ __rmw_create_node(
   const char * name,
   const char * namespace_)
 {
-  if (!name) {
-    RMW_SET_ERROR_MSG("name is null");
+  assert(nullptr != context);
+  assert(nullptr != context->impl);
+  assert(identifier == context->implementation_identifier);
+  static_cast<void>(identifier);
+
+  int validation_result = -1;
+  rmw_ret_t ret = rmw_validate_node_name(name, &validation_result, nullptr);
+  if (RMW_RET_OK != ret || RMW_NODE_NAME_VALID != validation_result) {
+    const char * reason = RMW_RET_OK == ret ?
+      rmw_node_name_validation_result_string(validation_result) :
+      rmw_get_error_string().str;
+    RMW_SET_ERROR_MSG_WITH_FORMAT_STRING("invalid node name: %s", reason);
     return nullptr;
   }
-
-  if (!namespace_) {
-    RMW_SET_ERROR_MSG("namespace_ is null");
+  validation_result = -1;
+  ret = rmw_validate_namespace(namespace_, &validation_result, nullptr);
+  if (RMW_RET_OK != ret || RMW_NAMESPACE_VALID != validation_result) {
+    const char * reason = RMW_RET_OK == ret ?
+      rmw_node_name_validation_result_string(validation_result) :
+      rmw_get_error_string().str;
+    RMW_SET_ERROR_MSG_WITH_FORMAT_STRING("invalid node namespace: %s", reason);
     return nullptr;
   }
 
-  rmw_node_t * node_handle = nullptr;
   auto common_context = static_cast<rmw_dds_common::Context *>(context->impl->common);
   rmw_dds_common::GraphCache & graph_cache = common_context->graph_cache;
-
-  node_handle = rmw_node_allocate();
-  if (!node_handle) {
-    RMW_SET_ERROR_MSG("failed to allocate rmw_node_t");
-    goto fail;
+  rcutils_allocator_t allocator = context->options.allocator;
+  rmw_node_t * node = reinterpret_cast<rmw_node_t *>(
+    allocator.zero_allocate(1u, sizeof(rmw_node_t), allocator.state));
+  if (nullptr == node) {
+    RMW_SET_ERROR_MSG("failed to allocate node");
+    return nullptr;
   }
-  node_handle->implementation_identifier = identifier;
-  node_handle->data = nullptr;
-
-  node_handle->name =
-    static_cast<const char *>(rmw_allocate(sizeof(char) * strlen(name) + 1));
-  if (!node_handle->name) {
-    RMW_SET_ERROR_MSG("failed to allocate memory");
-    node_handle->namespace_ = nullptr;  // to avoid free on uninitialized memory
-    goto fail;
+  auto cleanup_node = rcpputils::make_scope_exit(
+    [node, allocator]() {
+      allocator.deallocate(const_cast<char *>(node->name), allocator.state);
+      allocator.deallocate(const_cast<char *>(node->namespace_), allocator.state);
+      allocator.deallocate(node, allocator.state);
+    });
+  node->implementation_identifier = context->implementation_identifier;
```

Comment:
why not using the passed `identifier` argument?

Reply:
It's the same, though it seemed more correct to use the one in context. Actually, this function does not need to take a separate `identifier` but I chose not to change its signature.